_posts/2025-04-11-Hybrid-Search-Options-And-Practice.md

```diff
@@ -0,0 +1,275 @@
+
+
+
+#### Hybrid Search with Reciprocal Rank Fusion 
+
+Rank-based techniques focus on the position or rank of a document in the result set rather than its raw score. These methods are particularly useful when combining results from diverse sources where raw scores may not be directly comparable. RRF normalizes based on the rank of documents, making it robust to differences in scoring scales across different algorithms.
+
+
+```
+PUT /_search/pipeline/rrf-pipeline
+{
+  "description": "Post processor for hybrid RRF search",
+  "phase_results_processors": [
+    {
+      "score-ranker-processor": {
+        "combination": {
+          "technique": "rrf"
+        }
+      }
+    }
+  ]
+}
+```
+
+
+
+```
+GET /bookstore_catalog/_search?search_pipeline=rrf-pipeline
+{
+  "_source": {
+    "exclude": [
+      "title_embedding"
+    ]
+  },
+  "query": {
+    "hybrid": {
+      "queries": [
+        {
+          "match": {
+            "title": {
+              "query": "science fiction"
+            }
+          }
+        },
+        {
+          "neural": {
+            "title_embedding": {
+              "query_text": "science fiction",
+              "model_id": "aVeif4oB5Vm0Tdw8zYO2",
+              "k": 5
+            }
+          }
+        }
+      ]
+    }
+  }
+}
+```
+
+
+This query uses:
+
+* `hybrid`: Combines lexical and semantic searches
+* `search_pipeline`: Applies the RRF score combination
+
+
+How RRF works:
+
+
+1. **Sort documents by score**: Each query method sorts documents by score on every shard.
+2. **Assign rank positions**: Documents are ranked based on score for each query.
+3. **Apply the RRF formula**: The RRF score is computed using the following formula:
+
+```
+rankScore(document_i) = sum((1/(k + query_1_rank), (1/(k + query_2_rank), ..., (1/(k + query_j_rank)))
+
+```
+
+In this formula, `k` is a rank constant, and `query_j_rank` represents the ranking of a document in a particular query method. The example in the following diagram applies this formula using the default rank constant of 60.
+
+1. **Add rank contributions**: Rank calculations are combined, and documents are sorted by decreasing rank score.
+2. **Return the top results**: The highest-ranked documents are retrieved based on the query size.
+
+**When to use Rank Fusion:**
```

Comment:
One more item can be added - if nature of the query and/or data is changing over time, with fine-tuned weights that needs to be maintained continuously to keep relevancy high, with RRF this is not needed